vllm/engine/multiprocessing/mp_client.py

```diff
@@ -0,0 +1,389 @@
+
+    async def do_log_stats(self):
+        """Send a DO_LOG_STATS signal to the RPC Server"""
+        with suppress(MPClientClosedError):
+            await self._send_one_way_rpc_request(
+                request=RPCUtilityRequest.DO_LOG_STATS,
+                socket=self.input_socket)
+
+    @property
+    def is_running(self) -> bool:
+        return not self._errored
+
+    @property
+    def is_stopped(self) -> bool:
+        return self._errored
+
+    @property
+    def errored(self) -> bool:
+        return self._errored
+
+    async def generate(
+        self,
+        inputs: PromptInputs,
+        sampling_params: SamplingParams,
+        request_id: str,
+        lora_request: Optional[LoRARequest] = None,
+        trace_headers: Optional[Mapping[str, str]] = None,
+        prompt_adapter_request: Optional[PromptAdapterRequest] = None
+    ) -> AsyncGenerator[RequestOutput, None]:
+        """Send an RPCGenerateRequest to the RPCServer and stream responses."""
+
+        queue: asyncio.Queue[Union[RequestOutput,
+                                   BaseException]] = asyncio.Queue()
+        self.output_queues[request_id] = queue
+        finished = False
+        try:
+
+            # Send RPCGenerateRequest to the RPCServer.
+            await self.input_socket.send_multipart((cloudpickle.dumps(
+                RPCGenerateRequest(
+                    inputs=inputs,
+                    sampling_params=sampling_params,
+                    request_id=request_id,
+                    lora_request=lora_request,
+                    trace_headers=trace_headers,
+                    prompt_adapter_request=prompt_adapter_request)), ))
+
+            while not finished:
+                request_output = await queue.get()
+                if isinstance(request_output, BaseException):
+                    finished = True
+                    # On exception, check if the server is still healthy
+                    # possibly setting the `errored` property.
+                    if not self._errored:
+                        try:
+                            await self.check_health()
+                            pass
+                        except Exception as e:
+                            self._errored = True
+                            logger.exception(repr(e))
+                    raise request_output
+
+                finished = request_output.finished
+                yield request_output
+
+        finally:
+            self.output_queues.pop(request_id)
+            # Request was canceled by the client.
+            if not finished and not self._errored:
+                await self.abort(request_id)
+
+    async def check_health(self) -> None:
+        """Raise if unhealthy"""
+
+        await self._send_one_way_rpc_request(
+            request=RPCUtilityRequest.CHECK_HEALTH, socket=self.input_socket)
+
+        # Await acknowledgement from MPLLMEngine.
```

Comment:
I think this is a genuinely open question?

Reply:
It indeed seems not ideal. Either of the following approaches make more sense to me (but we probably don't have to fix it in this PR).
1. Make sure the response is corresponding to the right sender by matching request ID.
2. Cache the health check result in an interval (e.g., 10 seconds) and decouple client health check requests and the actual health checks.